trieste/acquisition/function.py

```diff
@@ -1121,3 +1123,238 @@ def penalization_function(x: TensorType) -> TensorType:
+    pending_points: Optional[TensorType] = None,
+    big_batch: bool = False,
+) -> AcquisitionFunction:
+    r"""
+    Return the General-purpose Information-Based Bayesian Optimization (GIBBON) acquisition function
+    of :cite:`Moss:2021`.The GIBBON acquisition function consists of two terms --- a quality term
+    and a diversity term. The quality term measures the amount of information that each individual
+    batch element provides about the objective function's minimal value :math:`y^*` (ensuring that
+    evaluations are targeted in promising areas of the space), whereas the repulsion term encourages
+    diversity within the batch (achieving high values for points with low predictive correlation).
+
+    GIBBON's repulsion term :math:`r=\log |C|`  is given by the log determinant of the predictive
+    correlation matrix :math:`C` between the `m` pending points and the current candidate.
+    The predictive covariance :math:`V` can be expressed as :math:V = [[v, A], [A, B]]` for a
+    tensor :math:`B` with shape [`m`,`m`] and so we can efficiently calculate :math:`|V|` using the
+    formula for the determinant of block matrices, i.e :math:`|V| = (v - A^T * B^{-1} * A) * |B|`.
+    Note that when using GIBBON for purely sequential optimization, the repulsion term is
+    not required.
+
+    As GIBBON's batches are built in a greedy manner, i.e sequentially adding points to build a set
+    of `m` pending points, we need only ever calculate the entropy reduction provided by adding the
+    current candidate point to the current pending points, not the full information gain provided by
+    evaluating all the pending points. This allows for a modest computational saving.
+
+    When performing BO with large batches (i.e. much larger than 10), GIBBON's approximations become
+    less accurate and its repulsion term dominates. Therefore, for large batch BO, we follow the
+    arguments of :cite:`Moss:2021` and divide GIBBON's repulsion term by :math:`B^{log B}`. This
+    behavior is accessed though setting `big_batch` to True.
+
+    :param model: The model of the objective function. GIBBON requires a model with
+        a :method:covariance_between_points method and so GIBBON only
+        supports :class:`GaussianProcessRegression` models.
+    :param samples: Samples from the distribution over :math:`y^*`.
+    :param pending_points: The points already chosen in the current batch.
+    :param big_batch: If True, then use the GIBBON formulation for BO with large batches.
+    :return: The GIBBON acquisition function. This function will raise :exc:`ValueError` or
+        :exc:`~tf.errors.InvalidArgumentError` if used with a batch size greater than one.
+    """
+    tf.debugging.assert_rank(samples, 2)
+    if len(samples) == 0:
+        raise ValueError("Min-value samples must be populated.")
+
+    try:
+        noise_variance = model.get_observation_noise()
+        model.covariance_between_points  # type: ignore
```

Comment:
Is the point of this to check whether covariance_between_points is defined? If so, I would expect this to raise an `AttributeError` on failure, not `NotImplementedError`. Also, it would probably be clearer to be to use `if hasattr(model, "get_observation_noise")` instead.